### ISSUE TITLE
DI: Nette\Http\Request subclass is required instead of interface

### ISSUE THREAD

**Issue by iinfo-dev-mk** [nette/http#90]
When we want to you own service for http request, subclass of Nette\Http\Request is required

``` php
// contex is instance of Nette\DI\Container
$context->removeService("http.request");
// OwnRequest implements Nette\Http\IRequest
$context->addService("http.request", new OwnRequest());
```

it throws exception

Nette\InvalidArgumentException: Service 'http.request' must be instance of Nette\Http\Request, OwnRequest given

I think same problem is with all services from Nette\Http package. Only interface can be requested, not class.

Maybe problem is with DI container itself, when _removeService_ is called, I wan't to remove all information about particular service, it seems that in DI container still remains some information about removed service, what isn't desirable.

**Comment by fprochazka** [comment]
There should be no references to `Nette\Http\Request` in Nette, since we've successfully replaced it with impl of it's interface and the application worked. You should first check your application, to make sure you're not requiring the class itself, instead of interface and then check the extensions you're using. Also maybe upgrade Nette, since the last referenced were removed few versions ago.

And also, the request object is a value object, that should in fact be a service. If you wanna work with multiple requests, you should use [Kdyby/RequestStack](https://github.com/Kdyby/RequestStack).

**Comment by iinfo-dev-mk** [comment]
@fprochazka I'm using [this default extension](https://github.com/nette/http/blob/v2.3/src/Bridges/HttpDI/HttpExtension.php). After container was created, I tried to replace http request services with code above, and it failed with the exception. 
It worked fine with nette 2.2, with nette 2.3 it failed.

**Comment by dg** [comment]
I remember that there was a break BC, when I changed it to interfaces, but I do not remember the details.

**Comment by matej21** [comment]
@dg yes, there would be a BC break if someone requires Http\Request directly and not the interface.

**Comment by dg** [comment]
Exactly.

**Comment by iinfo-dev-mk** [comment]
Require directly Http\Request directly isn't good practice if DIC offers service name by default. But replace particular service with another one is valid requirement for the DIC. 

@dg No, see the issue description, there is no issue with the configuration, I need to use removeService, addService calls. Replacing services this way doesn't work correctly. Hence I also noticed if problem isn't in DI container itself, see my description.

**Comment by dg** [comment]
I mean that you can change class in config.neon

```
services:
    http.request:
         class: Nette\Http\IRequest
```

and then replace it in bootstrap

**Comment by iinfo-dev-mk** [comment]
Again, no, I can't use any config file, there is only PHP code which wants to replace service in already built DIC ($context variable) ''dynamically'' via

```php
$context->removeService("http.request");
$context->addService("http.request", new OwnRequest())
```
It worked fine with nette 2.2 and fails with the exception in nette 2.3. 
There was always possibility change any service dynamically with removeService, addService, which doesn't work now. How can I change http request service dynamically?

Again, problem can be in DIC itself, when I call removeService(serviceName) I expect that DIC removes **all** information about particular service, but it seems that isn't true.

**Comment by dg** [comment]
Change it via `$configurator->addConfig(['services' => ['http.request' => ['class' => 'Nette\Http\IRequest']]])` or via extension.

Dynamic changes in container are limited, because it is statically compiled, so you cannot change one class to another when they are not descendants.

**Comment by iinfo-dev-mk** [comment]
I can't call any $configurator, PHP code knows only about  already built DIC, which is passed in.

If DIC can't dynamically replace service with instance of same interface such DIC is useless.
And secondly, current dynamic behaviour is inconsistent, I can replace other services for any instance of same interfaces dynamically without any problem. Finally, it was possible in nette 2.2 and you breaks it  silently in 2.3. 

Generally PHP is dynamic language, types are complete optional, any object can be passed as argument to any function and fails only in runtime if user want. So there is no reason to limit add, remove or replace services in DIC dynamically.

**Comment by matej21** [comment]
> If DIC can't dynamically replace service with instance of same interface such DIC is useless.

It can. Problem is that [there](https://github.com/nette/http/blob/master/src%2FBridges%2FHttpDI%2FHttpExtension.php#L38) is, for a compatibility, concrete class instead of the interface.

Why can't you modify config? If you are using IRequest everywhere, it won't break anything and you will be able to replace the service dynamically.

**Comment by iinfo-dev-mk** [comment]
@matej21 Our code with dynamic modification of DIC is used in special internal testing case from command line and only for that purpose there is necessary to replace default http request with special one, which is created also dynamically according command line arguments.
This is typical use case for dynamically behaviour of PHP and dynamically service change. 

I know that dynamic changing of DIC is the edge case, but shouldn't be limited to even interfaces. It's very handy to use some small piece of dynamic PHP code for such special situations, without modification of application itself, we are in PHP not in Java.

**Comment by milo** [comment]
@iinfo-dev-mk Noone argue, that DIC should allow to replace interface instance by another one. You are right. And DIC works in this way. But `Http\Request` is an exception, because it is added into DIC as a class, not as an interface. And it cannot be changed because of BC break.

If `OwnRequest` is used only for tests, you can workaround it.
```php
class OwnRequest2RequestWrapper extends Http\Request
{
    private $own;

    public function __construct(OwnRequest $own)
    {
        $this->own = $own;
    }

    # here overload all public methods to call $this->own methods
}
```

Such class instance can replace `Http\Requst` in existing DIC. And one day, you will delete `OwnRequest2RequestWrapper ` class and you will replace `IRequest` directly by `OwnRequst`.

It's not nice, it's workaround but IMHO it will work.

**Comment by dg** [comment]
> 
>  Finally, it was possible in nette 2.2 and you breaks it  silently in 2.3. 
> 

Yes, it is bad, can you please send pull request to documentation?

(Related commit 116eb121c897adf2dbc7a82f47dd367eb0196b11)

**Comment by iinfo-dev-mk** [comment]
@milo Yes I know, after all I eventually finish with inheriting from Http\Request but I don't like it, because it depends on Http\Request changes in the future and I need to investigate Http\Request and OwnRequest relation, it's much more fragile then before.

And still I think dynamic modifications of DIC via removeService, addService can be more relaxed, because it's primary targeted for special cases where such possibility is useful.

**Comment by iinfo-dev-mk** [comment]
@dg documentation - OK, I will try later this week

**Comment by dg** [comment]
This BC break can be probably solved via narrowed autowiring in DI 2.4 and alias.

**Comment by dg** [comment]
Fix reverted, solution is here https://github.com/nette/http/issues/90#issuecomment-214734843
